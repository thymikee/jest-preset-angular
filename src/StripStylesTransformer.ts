/*
 * Code is inspired by
 * https://github.com/kulshekhar/ts-jest/blob/25e1c63dd3797793b0f46fa52fdee580b46f66ae/src/transformers/hoist-jest.ts
 *
 * 
 * IMPLEMENTATION DETAILS:
 * This transformer handles one concern: removing styles.
 *
 * The property 'styles' inside a @Component(...) Decorator argument will
 * be modified, even if they are not used in the context of an
 * Angular Component.
 * 
 * This is the required AST to trigger the transformation:
 * 
 * ClassDeclaration
 *   Decorator
 *     CallExpression
 *       ObjectLiteralExpression
 *         PropertyAssignment
 *           Identifier
 *           StringLiteral
 */

// only import types, for the rest use injected `ConfigSet.compilerModule`
import {
  Node,
  SourceFile,
  TransformationContext,
  Transformer,
  Visitor,
  Identifier,
  ClassDeclaration,
  PropertyAssignment
} from 'typescript';
import { ConfigSet } from './TransformUtils';

/** Angular component decorator Styles property name */
const STYLES = 'styles';
/** Angular component decorator name */
const COMPONENT = 'Component';
/** All props to be transformed inside a decorator */
const TRANSFORM_IN_DECORATOR_PROPS = [STYLES];

/**
 * Transformer ID
 * @internal
 */
export const name = 'angular-component-strip-styles';

// increment this each time the code is modified
/**
 * Transformer Version
 * @internal
 */
export const version = 1;

/**
 * The factory of hoisting transformer factory
 * @internal
 */
export function factory(cs: ConfigSet) {
  /**
   * Our compiler (typescript, or a module with typescript-like interface)
   */
  const ts = cs.compilerModule;

  /**
   * Traverses the AST down inside a decorator to a styles assignment
   * and returns a boolean indicating if it should be transformed.
   */
  function isInDecoratorPropertyAssignmentToTransform(node: Node): node is ClassDeclaration {
    return getInDecoratorPropertyAssignmentsToTransform(node).length > 0;
  }

  /**
   * Traverses the AST down inside a decorator to a styles assignment
   * returns it in an array.
   */
  function getInDecoratorPropertyAssignmentsToTransform(node: Node) {
    if (!ts.isClassDeclaration(node) || !node.decorators) {
      return [];
    }

    return node.decorators
      .map(dec => dec.expression)
      .filter(ts.isCallExpression)
      .filter(callExpr => ts.isCallExpression(callExpr) && ts.isIdentifier(callExpr.expression) && callExpr.expression.getText() === COMPONENT)
      .reduce((acc, nxtCallExpr) => Array.prototype.concat.apply(acc, 
        nxtCallExpr.arguments
          .filter(ts.isObjectLiteralExpression)
          .reduce((acc, nxtArg) => Array.prototype.concat.apply(acc,
            nxtArg.properties
              .filter(ts.isPropertyAssignment)
              .filter(propAss =>
                ts.isIdentifier(propAss.name) &&
                TRANSFORM_IN_DECORATOR_PROPS.includes(
                  (propAss.name as Identifier).text
                )
              )
          ), [] as PropertyAssignment[])
      ), [] as PropertyAssignment[])
  }

  /**
   * Clones the styles assignment and manipulates it.
   * @param node the property assignment to change
   */
  function transfromStylesAssignmentForJest(node: ClassDeclaration) {
    const mutableNode = ts.getMutableClone(node)
    const assignments = getInDecoratorPropertyAssignmentsToTransform(mutableNode)

    assignments.forEach(assignment => {
      if ((assignment.name as Identifier).text === STYLES) {
        // replace initializer array with empty array
        assignment.initializer = ts.createArrayLiteral()
      }
    })
    return mutableNode
  }

  /**
   * Create a source file visitor which will visit all nodes in a source file
   * @param ctx The typescript transformation context
   * @param _ The owning source file
   */
  function createVisitor(ctx: TransformationContext, _: SourceFile) {
    /**
     * Main visitor, which will be called recursively for each node in the source file's AST
     * @param node The node to be visited
     */
    const visitor: Visitor = node => {
      // before we create a deep clone to modify, we make sure that
      // this is an assignment which we want to transform
      if (isInDecoratorPropertyAssignmentToTransform(node)) {
        // get transformed node with changed properties
        return transfromStylesAssignmentForJest(node);
      } else {
        // else look for assignments inside this node recursively
        return ts.visitEachChild(node, visitor, ctx);
      }
    };
    return visitor;
  }

  return (ctx: TransformationContext): Transformer<SourceFile> => (
    sf: SourceFile
  ) => ts.visitNode(sf, createVisitor(ctx, sf));
}
