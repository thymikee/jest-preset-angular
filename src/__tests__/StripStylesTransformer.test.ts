/*
 * Code is inspired by
 * https://github.com/kulshekhar/ts-jest/blob/25e1c63dd3797793b0f46fa52fdee580b46f66ae/src/transformers/hoist-jest.spec.ts
 *
 */

import * as tsc from 'typescript';
import * as transformer from '../StripStylesTransformer';

const CODE_WITH_STYLES_AND_OTHER_ASSIGNMENTS = `
  import { Component } from '@angular/core';

  @SomeDecorator({
    value: 'test',
    styles: [
      ':host { background-color: red }'
    ],
  })
  @Component({
    templateUrl: './page.html',
    styleUrls: [
      './fancy-styles.css',
      './basic-styles.scss'
    ],
    styles: [
      'body { display: none }',
      'html { background-color: red }'
    ],
    unaffectedProperty: 'whatever'
  })
  export class AngularComponent {
  }
`;

const CODE_WITH_ASSIGNMENT_OUTSIDE_DECORATOR = `
  const assignmentsToNotBeTransformed = {
    styles: [{
      color: 'red'
    }]
  };
`;

const createFactory = () => {
  return transformer.factory({ compilerModule: tsc } as any);
};
const transpile = (source: string) =>
  tsc.transpileModule(source, { transformers: { before: [createFactory()] } });

describe('inlining template and stripping styles', () => {
  it('should not strip styleUrls assignment', () => {
    const out = transpile(CODE_WITH_STYLES_AND_OTHER_ASSIGNMENTS);

    expect(out.outputText).toMatchSnapshot();
  });

  it('should not transform styles outside decorator', () => {
    const out = transpile(CODE_WITH_ASSIGNMENT_OUTSIDE_DECORATOR);

    expect(out.outputText).toMatchSnapshot();
  });
});
