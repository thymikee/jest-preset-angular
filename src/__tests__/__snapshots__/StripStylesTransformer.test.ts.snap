// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inlining template and stripping styles should not transform styles outside decorator 1`] = `
"var assignmentsToNotBeTransformed = {
    styles: [{
            color: 'red'
        }]
};
"
`;

exports[`inlining template and stripping styles should not strip styleUrls assignment 1`] = `
"\\"use strict\\";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === \\"object\\" && typeof Reflect.decorate === \\"function\\") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var core_1 = require(\\"@angular/core\\");
var AngularComponent = /** @class */ (function () {
    function AngularComponent() {
    }
    AngularComponent = __decorate([
        SomeDecorator({
            value: 'test',
            styles: [
                ':host { background-color: red }'
            ],
        }),
        core_1.Component({
            templateUrl: './page.html',
            styleUrls: [
                './fancy-styles.css',
                './basic-styles.scss'
            ],
            styles: [],
            unaffectedProperty: 'whatever'
        })
    ], AngularComponent);
    return AngularComponent;
}());
exports.AngularComponent = AngularComponent;
"
`;
